services/src/main/java/org/keycloak/services/HttpResponseImpl.java

```diff
@@ -44,6 +56,31 @@ public void setHeader(String name, String value) {
         delegate.getOutputHeaders().putSingle(name, value);
     }
 
+    @Override
+    public void addCookieIfAbsent(HttpCookie cookie) {
+        if (cookie == null) {
+            throw new IllegalArgumentException("Cookie is null");
+        }
+
+        if (cookies == null) {
+            cookies = new HashSet<>();
+        }
+
+        if (cookies.add(cookie)) {
```

Comment:
This relies on hasCode in Cookie:

```
    public int hashCode() {
        int hash = 7;
        hash = 97 * hash + (this.name != null ? this.name.hashCode() : 0);
        hash = 97 * hash + (this.value != null ? this.value.hashCode() : 0);
        hash = 97 * hash + this.version;
        hash = 97 * hash + (this.path != null ? this.path.hashCode() : 0);
        hash = 97 * hash + (this.domain != null ? this.domain.hashCode() : 0);
        return hash;
    }
```

So would be possible to have two set-cookie for the same cookie, but with different values. Would it be better to make cookies a list and search for existing cookies by name/path/domain instead?

Reply:
Maybe I'm overthinking things a bit, wdyt? just keep as is, or update to compare based on name/path/domain as I suggested?

Reply:
I would keep it as is for now and rely on JAX-RS API as much as possible.